Optional OomErrors

Mixthos opened the pull request:
(Mostly) as discussed in #20.
`SwapchainImage::from_raw` isn't changed yet.
I fixed the teapot example to work with my changes. The other examples were broken anyway.
I'm not confident about my ability to change stuff inside these huge macros without breaking them, so you should probably double-check them. ;)

tomaka commented on vulkano/src/image/sys.rs:
```diff
@@ -479,6 +479,12 @@ impl UnsafeImage {
     }
 
     #[inline]
+    pub unsafe fn bind_memory(&self, memory: &DeviceMemory, offset: usize)
```
`bind_memory` doesn't need a `raw` variant. It's an unsafe and dangerous function which isn't supposed to be used directly, so making it more convenient to use doesn't make sense.

Mixthos replied:
I reverted it.

tomaka commented on vulkano/src/framebuffer/sys.rs:
```diff
@@ -52,7 +52,7 @@ impl UnsafeRenderPass {
     /// mode.
     ///
-    pub unsafe fn new<Ia, Ip, Id>(device: &Arc<Device>, attachments: Ia, passes: Ip,
+    pub unsafe fn raw<Ia, Ip, Id>(device: &Arc<Device>, attachments: Ia, passes: Ip,
```
Since the documentation is so long, could you remove the docs from `raw` and just say `See the docs of new()`?

tomaka commented:
Thanks! That's very nice!

About your comments on the fact that Arcs can be omitted, I'm really not sure. It depends on how semaphores/fences are going to be pooled, whose design is currently blocked by some ambiguities in the Vulkan specs.

Did you try running `cargo test --lib` to see if all the tests compile?

Mixthos commented:
> Did you try running cargo test --lib to see if all the tests compile?

Oops! I knew I forgot _something_. I'll fix those now.

Mixthos commented:
> bind_memory doesn't need a raw variant. It's an unsafe and dangerous function which isn't supposed to be used directly, so making it more convenient to use doesn't make sense.

I reverted it in cebabe1f639c45037cb8d40d2a5ca3a88ecba722.

> Since the documentation is so long, could you remove the docs from raw and just say See the docs of new()?
> This would avoid having to maintain the same docs twice.

I replaced all docs for the "raw" functions in b8269ead29d1b30af09ed8eaddea1095b62f1183.

> Did you try running cargo test --lib to see if all the tests compile?

I fixed them in b6b073c888dbb257d2c904572aabfb5d2e69a148:

> test result: ok. 64 passed; 0 failed; 6 ignored; 0 measured

Did I forget anything?

tomaka commented:
Thanks!
